### PR TITLE
Move connections pane into core positron

### PR DESCRIPTION
**Pull request by dfalbel**
Addresses #4874 

Needs codicons from: https://github.com/posit-dev/positron-codicons/pull/6
Needs ark side: https://github.com/posit-dev/ark/pull/589

This PR is an initial implementation of the connections pane into Positron core. It's still missing a lot of functionality, but will eventually be equivalent to the current positron-connections extension.

TODO:

- [x] Workspace storage of previous connections
- [x] Setting to opt-in for the new implementation
- [x] Button to open data explorer for a `Table` or `View`
- [x] Codicons for most common DB data types (schema, collections, etc) and for creating a new connection
- [x] Review event listeners disposing
- [x] Implement Python side of new GetMetadata RPC
- [x] Implement refresh connections
- [x] Display the field type 
- [x] Display the language for each connection
- [x] Review error handling
- [x] Fix smoke tests (smoke tests are passing locally)
- [x] Fix integration tests
- [x] Rename 'Connections Core' to just 'Connections'
- [x] Support for the focus event

Demo:

https://github.com/user-attachments/assets/9e7b7905-3c1b-421d-a9e4-6c8e5d3b3b94

Must be enabled with:

image

### CODE REVIEW

File: src/vs/workbench/contrib/positronConnections/browser/positronConnections.tsx [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+	const [height, setHeight] = React.useState(props.reactComponentContainer.height);
+
+	useEffect(() => {
+		const disposableStore = new DisposableStore();
+		disposableStore.add(props.reactComponentContainer.onSizeChanged(size => {
+			setWidth(size.width);
+			setHeight(size.height);
+		}));
+		return () => disposableStore.dispose();
+	}, []);
```
Comment (softwarenerd) [inline src/vs/workbench/contrib/positronConnections/browser/positronConnections.tsx:51]: Missing dependencies:

```
	useEffect(() => {
		const disposableStore = new DisposableStore();
		disposableStore.add(props.reactComponentContainer.onSizeChanged(size => {
			setWidth(size.width);
			setHeight(size.height);
		}));
		return () => disposableStore.dispose();
	}, [props.reactComponentContainer]);
```

File: src/vs/workbench/contrib/positronConnections/browser/positronConnections.tsx [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+		disposableStore.add(props.reactComponentContainer.onRestoreScrollPosition(() => {
+			if (scrollStateRef.current) {
+				if (innerRef.current) {
+					DOM.restoreParentsScrollTop(innerRef.current, scrollStateRef.current);
+				}
+				setScrollState(undefined);
+			}
+		}));
+		return () => disposableStore.dispose();
+	}, []);
```
Comment (softwarenerd) [inline src/vs/workbench/contrib/positronConnections/browser/positronConnections.tsx:73]: Missing dependencies:

```
	useEffect(() => {
		const disposableStore = new DisposableStore();
		disposableStore.add(props.reactComponentContainer.onSaveScrollPosition(() => {
			if (innerRef.current) {
				setScrollState(DOM.saveParentsScrollTop(innerRef.current));
			}
		}));
		disposableStore.add(props.reactComponentContainer.onRestoreScrollPosition(() => {
			if (scrollStateRef.current) {
				if (innerRef.current) {
					DOM.restoreParentsScrollTop(innerRef.current, scrollStateRef.current);
				}
				setScrollState(undefined);
			}
		}));
		return () => disposableStore.dispose();
	}, [props.reactComponentContainer, scrollStateRef, setScrollState]);
```

File: src/vs/workbench/contrib/positronConnections/browser/positronConnections.tsx [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+	const [items, setItems] = useState<IPositronConnectionEntry[]>(props.connectionsService.getConnectionEntries);
+	useEffect(() => {
+		const disposableStore = new DisposableStore();
+		disposableStore.add(props.connectionsService.onDidChangeEntries((entries) => {
+			setItems(entries);
+		}));
+		// First entries refresh - on component mount.
+		props.connectionsService.refreshConnectionEntries();
+		return () => disposableStore.dispose();
+	}, []);
```
Comment (softwarenerd) [inline src/vs/workbench/contrib/positronConnections/browser/positronConnections.tsx:84]: Missing dependencies:

```
	useEffect(() => {
		const disposableStore = new DisposableStore();
		disposableStore.add(props.connectionsService.onDidChangeEntries((entries) => {
			setItems(entries);
		}));
		// First entries refresh - on component mount.
		props.connectionsService.refreshConnectionEntries();
		return () => disposableStore.dispose();
	}, [props.connectionsService]);
```

File: src/vs/workbench/contrib/positronConnections/browser/components/actionBar.tsx [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	clearAllHandler: () => void;
+}
+
+export const ActionBar = (props: React.PropsWithChildren<ConnectionActionBarProps>) => {
+
+	// We only enable the disconnect button if:
+	// 1. there's some connection selected
+	// 2. it's the root of a connection (level == 0).
+	// 3. the connection is active.
+	const disconnectDisabled = (props.selectedEntry === undefined) ||
```
Comment (softwarenerd) [inline src/vs/workbench/contrib/positronConnections/browser/components/actionBar.tsx:49]: NIT: All the extra parenthesis make me wonder?

File: src/vs/workbench/contrib/positronConnections/browser/components/actionBar.tsx [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+		(props.selectedEntry.level !== 0) ||
+		(!props.selectedEntry.active);
+
+	// We only enable the connect button if:
+	// 1. there's some connection selected
+	// 2. it's the root of a connection (level == 0).
+	// 3. the connection is not active.
+	// 4. it implements a 'connect' method.
+	const connectDisabled = (props.selectedEntry === undefined) ||
```
Comment (softwarenerd) [inline src/vs/workbench/contrib/positronConnections/browser/components/actionBar.tsx:58]: NIT: All the extra parenthesis make me wonder?

File: src/vs/workbench/contrib/positronConnections/browser/components/actionBar.tsx [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+import { IKeybindingService } from 'vs/platform/keybinding/common/keybinding';
+import { ActionBarSearch } from 'vs/platform/positronActionBar/browser/components/actionBarSearch';
+
+import 'vs/css!./actionBar';
+import { IPositronConnectionEntry } from 'vs/workbench/services/positronConnections/browser/positronConnectionsCache';
+
+const kPaddingLeft = 8;
+const kPaddingRight = 8;
+export const kHeight = 32;
+
```
Comment (softwarenerd) [inline src/vs/workbench/contrib/positronConnections/browser/components/actionBar.tsx:28]: NIT: In general, in VS Code, `const` are declared like:

```
// Activating extension for Multiroot and Debugger CI tests for Windows takes just over 2 minutes sometimes, so 3 minutes seems like a safe margin
export const MAX_EXTENSION_ACTIVATION_TIME = 180_000;
export const TEST_TIMEOUT = 60_000;
export const TEST_RETRYCOUNT = 3;
export const IS_SMOKE_TEST = process.env.VSC_PYTHON_SMOKE_TEST === '1';
export const IS_PERF_TEST = process.env.VSC_PYTHON_PERF_TEST === '1';
export const IS_MULTI_ROOT_TEST = isMultitrootTest();

// If running on CI server, then run debugger tests ONLY if the corresponding flag is enabled.
export const TEST_DEBUGGER = IS_CI_SERVER ? IS_CI_SERVER_TEST_DEBUGGER : true;
```

File: src/vs/workbench/contrib/positronConnections/browser/positronConnections.tsx [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+
+	const [items, setItems] = useState<IPositronConnectionEntry[]>(props.connectionsService.getConnectionEntries);
+	useEffect(() => {
+		const disposableStore = new DisposableStore();
+		disposableStore.add(props.connectionsService.onDidChangeEntries((entries) => {
+			setItems(entries);
+		}));
+		// First entries refresh - on component mount.
+		props.connectionsService.refreshConnectionEntries();
```
Comment (softwarenerd) [inline src/vs/workbench/contrib/positronConnections/browser/positronConnections.tsx:82]: `refreshConnectionEntries` is not awaited? Makes me wonder if that's actually OK.

File: src/vs/workbench/contrib/positronConnections/browser/positronConnections.tsx [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+		return () => disposableStore.dispose();
+	}, []);
+
+	const [selectedId, setSelectedId] = useState<string>();
+
+	const ItemEntry = (props: ItemEntryProps) => {
+		const itemProps = items[props.index];
+
+		return (
+			<PositronConnectionsItem
```
Comment (softwarenerd) [inline src/vs/workbench/contrib/positronConnections/browser/positronConnections.tsx:92]: Perhaps:

```
		return (
			<PositronConnectionsItem
				{...itemProps}
				selected={itemProps.id === selectedId}
				onSelectedHandler={() => setSelectedId(itemProps.id)}
				style={props.style}>
			</PositronConnectionsItem>
		);
```

Just a thought.

**Review by softwarenerd — DISMISSED** [review]
Some nits. A couple of thoughts and a few questions.

File: src/vs/workbench/contrib/positronConnections/browser/positronConnections.tsx (hunk 8 above)
Comment (dfalbel) [reply src/vs/workbench/contrib/positronConnections/browser/positronConnections.tsx:92]: I tried this and it doesn't work. I have no idea why :S
I just changed to https://github.com/posit-dev/positron/pull/4967/commits/c34d6c7118ef540d36288da17cc666802221ae34 which looks slightly better, I think!

**Pull request merged**